src/server/controllers/UserController.ts

```diff
@@ -1,185 +1,132 @@
-      hours: asString,
-      enabled: asBoolean
-    })
-
-    const { userId } = asQuery(req.query)
-    const { currencyCode } = asParams(req.params)
-    const { hours, enabled } = asBody(req.body)
-
-    const user = await User.fetch(userId)
-    const currencySettings = user.notifications.currencyCodes[currencyCode]
-    // @ts-expect-error
-    currencySettings[hours] = enabled
-    await user.save()
-
-    console.log(
-      `Updated notification settings for user ${userId} for ${currencyCode}`
-    )
-
-    res.json(currencySettings)
-  } catch (err) {
-    console.error(
-      `Failed to update notification settings for user ${String(
-        req.query.userId
-      )} for ${req.params.currencyCode}`,
-      err
-    )
-    res.status(500).json(err)
-  }
-})
-
-UserController.post('/notifications/toggle', async (req, res) => {
-  try {
-    const asQuery = asObject({
-      userId: asString
-    })
-    const asBody = asObject({
-      enabled: asBoolean
-    })
-    console.log(req.body)
-
-    const { userId } = asQuery(req.query)
-    const { enabled } = asBody(req.body)
-
-    let user = await User.fetch(userId)
-    if (!user) user = new User(null, userId)
-    user.notifications.enabled = enabled
-    await user.save()
-
-    console.log(`User notifications toggled to: ${String(enabled)}`)
-
-    res.json(user)
-  } catch (err) {
-    console.error(`Failed to toggle user notifications`, err)
-    res.status(500).json(err)
-  }
-})
+export const getUser: Serverlet<ApiRequest> = async request => {
+  const { log, query } = request
+
+  const asQuery = asObject({
+    userId: asString
+  })
+
+  const { userId } = asQuery(query)
+  const result = await User.fetch(userId)
+
+  log(`Got user settings for ${userId}`)
+
+  return jsonResponse(result)
+}
+
+export const postUserDevice: Serverlet<ApiRequest> = async request => {
+  const { log, query } = request
+
+  const asQuery = asObject({
+    deviceId: asString,
+    userId: asString
+  })
+
+  const { deviceId, userId } = asQuery(query)
+
+  let user = await User.fetch(userId)
+  if (!user) user = new User(null, userId)
+
+  await user.attachDevice(deviceId)
+
+  log(`Successfully attached device "${deviceId}" to user "${userId}"`)
+
+  return jsonResponse(user)
+}
+
+export const postUserNotifications: Serverlet<ApiRequest> = async request => {
+  const { log, json, query } = request
+
+  const asQuery = asObject({
+    userId: asString
+  })
+  const asBody = asObject({
+    currencyCodes: asArray(asString)
+  })
+
+  const { userId } = asQuery(query)
+  const { currencyCodes } = asBody(json)
+
+  const user = await User.fetch(userId)
+  await user.registerNotifications(currencyCodes)
+
+  log(`Registered notifications for user ${userId}: ${String(currencyCodes)}`)
+
+  return jsonResponse(user)
+}
+
+export const getUserCurrency: Serverlet<ApiRequest> = async request => {
+  const { log, path, query } = request
+  const match = path.match(/notifications\/([0-9A-Za-z]+)\/?$/)
```

Comment:
I assume you'd have an url path like "/v1/user/notifications/BTC/userID?=123456" ?

In this case, the regex will return null. It will only work on things like "/v1/user/notifications/BTC/", but I assume if we have some query parameters appended to the path, the regex won't work.

Reply:
It works because Express removes the `?userId=xyz` part of the path and shoves it in the `query`, so the regex doesn't need to deal with it.